### PR TITLE
Add more info in trace-api-plugin - 2.0

### PR DESCRIPTION
**Pull request by oldcold**
## Change Description

Add more information in `trace-api-plugin` responses for better usage.
- Add `transaction_mroot`, `action_mroot` and `schedule_version` in block trace;
- Add `status`, `cpu_usage_us`, `net_usage_words`, `signatures` and `transaction_header` in transaction trace.  

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [x] Documentation Additions

See additional data included above.

### CODE REVIEW

File: plugins/trace_api_plugin/test/test_extraction.cpp [hunk 1]
```diff
@@ -228,37 +265,54 @@ BOOST_AUTO_TEST_SUITE(block_extraction)
...
                ptrx2.id(),
                {
                   {
                      1,
-                     "bob"_n, "eosio.token"_n, "transfer"_n,
-                     {{ "bob"_n, "active"_n }},
-                     make_transfer_data( "bob"_n, "alice"_n, "0.0001 SYS"_t, "Memo!" )
+                     "fred"_n, "eosio.token"_n, "transfer"_n,
+                     {{ "fred"_n, "active"_n }},
+                     make_transfer_data( "fred"_n, "bob"_n, "0.0001 SYS"_t, "Memo!" )
```
Comment (heifner) [inline plugins/trace_api_plugin/test/test_extraction.cpp:283]: Why did the order change?

Comment (oldcold) [reply plugins/trace_api_plugin/test/test_extraction.cpp]: Used when testing, rollbacked.

File: plugins/trace_api_plugin/include/eosio/trace_api/extract_util.hpp [hunk 2]
```diff
@@ -39,6 +39,37 @@ inline transaction_trace_v0 to_transaction_trace_v0( const chain::transaction_tr
    return r;
 }
 
+inline transaction_trace_v1 to_transaction_trace_v1( const chain::transaction_trace_ptr& t, const chain::signed_transaction& trx = {} ) {
+    transaction_trace_v1 r;
+    if( !t->failed_dtrx_trace ) {
+        r.id = t->id;
+        r.status = t->receipt->status;
```
[the remark this answers is not in the document]
Comment (oldcold) [reply plugins/trace_api_plugin/include/eosio/trace_api/extract_util.hpp]: 1. `status` is included for the purpose of caching failed scheduled transactions if any, which might be desired by end-users.  
2. We were trying to record actual CPU/NET usage of both executed and failed transactions, information from top level trace should do the trick.

File: plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp [hunk 3]
```diff
@@ -20,25 +21,62 @@ namespace eosio { namespace trace_api {
...
-   };
+  struct transaction_trace_v1 : public transaction_trace_v0 {
+
+     transaction_trace_v1(chain::transaction_id_type id, std::vector<action_trace_v0> actions, fc::enum_type<uint8_t,status_type> status, uint32_t cpu_usage_us, fc::unsigned_int net_usage_words, std::vector<chain::signature_type> signatures, chain::transaction_header trx_header)
+     : transaction_trace_v0{id, std::move(actions)}
+     , status(status)
+     , cpu_usage_us(cpu_usage_us)
+     , net_usage_words(net_usage_words)
+     , signatures(std::move(signatures))
+     , trx_header(trx_header) {}
```
[the remark this answers is not in the document]
Comment (oldcold) [reply plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp:39]: Used in unit tests.

File: plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp [hunk 4]
```diff
@@ -20,25 +21,62 @@ namespace eosio { namespace trace_api {
...
+
+     transaction_trace_v1() {}
```
[the remark this answers is not in the document]
Comment (oldcold) [reply plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp]: Used in unit tests.

File: plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp [hunk 5]
```diff
@@ -20,25 +21,62 @@ namespace eosio { namespace trace_api {
...
+  };
+
+  struct block_trace_v1 : public block_trace_v0 {
+
+     block_trace_v1(chain::block_id_type id, uint32_t number, chain::block_id_type previous_id, chain::block_timestamp_type timestamp, chain::name producer, chain::checksum256_type transaction_mroot, chain::checksum256_type action_mroot, uint32_t schedule_version, std::vector<transaction_trace_v1> transactions_v1)
+     : block_trace_v0{id, number, previous_id, timestamp, producer, {}}
+     , transaction_mroot(transaction_mroot)
+     , action_mroot(action_mroot)
+     , schedule_version(schedule_version)
+     , transactions_v1(std::move(transactions_v1)) {}
```
[the remark this answers is not in the document]
Comment (oldcold) [reply plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp:66]: Used in unit tests.

File: plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp [hunk 6]
```diff
@@ -20,25 +21,62 @@ namespace eosio { namespace trace_api {
...
+
+     block_trace_v1() {}
```
[the remark this answers is not in the document]
Comment (oldcold) [reply plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp]: Used in unit tests.

File: plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp:39]: Not required. The tests will work without it, like they did before.

File: plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/trace_api_plugin/include/eosio/trace_api/trace.hpp:66]: Not required. The tests will work without it, like they did before.

File: plugins/trace_api_plugin/include/eosio/trace_api/extract_util.hpp [hunk 7]
```diff
@@ -39,6 +39,27 @@ inline transaction_trace_v0 to_transaction_trace_v0( const chain::transaction_tr
...
+inline transaction_trace_v1 to_transaction_trace_v1( const cache_trace& t ) {
+    transaction_trace_v1 r;
+    if( !t.trace->failed_dtrx_trace ) {
+        r.id = t.trace->id;
+    } else {
+        r.id = t.trace->failed_dtrx_trace->id; // report the failed trx id since that is the id known to user
+    }
+    r.status = t.trace->receipt->status;
+    r.cpu_usage_us = t.trace->receipt->cpu_usage_us;
+    r.net_usage_words = t.trace->receipt->net_usage_words;
```
Comment (heifner) [inline plugins/trace_api_plugin/include/eosio/trace_api/extract_util.hpp:51]: As is `to_transaction_trace_v1` can't be called with an empty `optional t.trace->receipt` however, with the addition of `status` I can see someone adding failed in the future. Add a check that `receipt` exists.

**Comment by heifner** [comment]
Please create a version of this targeted to `develop`, we will then merge them at the same time.

**Pull request merged**